python/parquet_tools/csv2pq_schema.py

```diff
@@ -0,0 +1,260 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+This module encapsulates all the code to process schema files and provide
+support for schemas that allow null values for non-floating point values.
+
+This module augments schemas that allow null values for data types which
+pandas does not support to be null. Specifically, for each such column is
+adds a  column with a name of '<cname>_ISNULL', where <cname> is the name
+of the column that may contain an unsupported null value. The associated
+column will contain true if the original column contains a null value or
+false, otherwise. Such a column will contain a well defined value or a
+value specified by the command issuer. While this module sets up the
+mechanics, the relealiztion of this process is handled by csv2pq_handler.py.
+"""
+
+__all__ = ['chkSchema', 'getSchema']
+
+import csv
+import sys
+import numpy as np
+
+from csv2pq_cmdinfo import CmdInfo
+from csv2pq_utils import Fatal
+
+
+# *****************************************************************************
+# *           T y p e   C o n v e r s i o n   I n f o r m a t i o n           *
+# *****************************************************************************
+
+class TypeInfo:
+
+    table = {'int':     np.int32,   'short':    np.int16, 'long':  np.int64,
+             'bigint':  np.int64,   'smallint': np.int16,
+             'int8':    np.int8,    'int32':    np.int32, 'int64': np.int64,
+             'float':   np.float32, 'double':   np.float64,
+             'float32': np.float32, 'float64':  np.float64,
+             'char':    np.bytes_,  'bool':     np.bool_
+             }
+
+    isInt = {'int':     True,       'short':    True,     'long':  True,
+             'bigint':  True,       'smallint': True,
+             'int8':    True,       'int32':    True,     'int64': True
+             }
+
+
+# *****************************************************************************
+# *                                a d d C o l                                *
+# *****************************************************************************
+
+def addCol(cName):
+    "Add a column to the column list and record size of longest name."
+
+    CmdInfo.colNames.append(cName)
+    if len(cName) > CmdInfo.cNameMax: CmdInfo.cNameMax = len(cName)
+
+
+# *****************************************************************************
+# *                               g e t D e c T                               *
+# *****************************************************************************
+
+def getDecT(dArgs):
+    "Convert decimal type to appropriate numpy type."
```

Comment:
Ah, you need to further explain your request, please.

Reply:
I mean what format of input string it accepts, without some description of what it takes and what it returns it's hard for me to validate that behavior is correct. And after you done that, it would be  also nice to have a unit test for this method 🙂

Reply:
There is only one problem unit testing this very low level function, it is not public. Are you suggesting I make it public and violate other rules? Obviously, the only other way of unit testing this is to run an acual schema through it with all possible variation, capturing the output and seeing it it's consistent. That's sure a lot of work for a tiny function.

Reply:
Unit tests, as part of the package, do not have to use public interfaces.